### PR TITLE
v2.1: Use PMIx_Log() for show_help() messages. 

### PR DESCRIPTION
**Pull request by awlauria**
This moves a lot of dead code in show_help.c to OpenPMIx,
which will now do the de-duping and aggregation for us
via the PMIx_Log() function.

Otherwise the other big change is that the directives
are now packaged up as well as the infos. Othwewise
when prrte is forwarding the logging messages to the hnp,
it will miss out on any aggregation requests.

PMIx side change (master): https://github.com/openpmix/openpmix/pull/2569

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>
(cherry picked from commit https://github.com/openpmix/prrte/commit/b80d1311ca55549185ca00b54510e00fb000d39f)

[Control aggregation behavior](https://github.com/openpmix/prrte/pull/1354/commits/e13babbe90c4c294c714fcb14a192a196d1edcc6) 

Define a cmd line option to disable aggregation of show_help messages,
defaulting to enabling de-aggregation.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>
Partially-cherry-picked-from: https://github.com/openpmix/prrte/commit/47b5ad1653123e984a644ef84b9a7aaf4e9b5fb3
Co-authored-by: Ralph Castain <rhc@pmix.org>

[Restore noloop for logging.](https://github.com/openpmix/prrte/commit/382594804af8c54ab331369f072d19110373b7ff) 

This seems to have been accidently changed in the logging
update.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>
(cherry picked from commit https://github.com/openpmix/prrte/commit/da12b9da13792655e9f9d60c620040ecea83f88e)

[Protect against compiling with PMIx without agg support.](https://github.com/openpmix/prrte/pull/1354/commits/950672ee9bf53e283e8ff72bd15a6cc5c3e73925) 

We can still log to stderr in this case, will just
lost aggregate and de-dup support.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

[Remove setting of PRTE_MCA_prte_base_help_aggregate.](https://github.com/openpmix/prrte/pull/1354/commits/d870390f2d8493c6ec08f982bf99671f17473c3c) 

This snuck through here: https://github.com/openpmix/prrte/pull/1350

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>
(cherry picked from commit https://github.com/openpmix/prrte/commit/8103bff7880e68398f5c0d4c654c08278911d44d)

Refs: https://github.com/openpmix/openpmix/pull/2576

### CODE REVIEW

File: src/tools/prte/prte.c [hunk 1]
```diff
@@ -463,8 +463,9 @@ int main(int argc, char *argv[])
     if (NULL != opt) {
         prte_pmix_server_globals.report_uri = strdup(opt->values[0]);
     }
-    /* don't aggregate help messages as that will apply job-to-job */
-    prte_setenv("PRTE_MCA_prte_base_help_aggregate", "0", true, &environ);
+
+    /* aggregate messages from prte. */
+    prte_setenv("PRTE_MCA_prte_base_help_aggregate", "1", true, &environ);
```
Comment (rhc54) [inline src/tools/prte/prte.c:468]: Err...we just removed that MCA param, so why set it here? It is "on" by default for each job anyway.

File: src/tools/prted/prted.c [hunk 2]
```diff
@@ -344,6 +344,8 @@ int main(int argc, char *argv[])
     /* ensure we silence any compression warnings */
     prte_setenv("PMIX_MCA_compress_base_silence_warning", "1", true, &environ);
 
+    /* Aggregate any show helps from prted. */
+    pmix_setenv("PRTE_MCA_prte_base_help_aggregate", "1", true, &environ);
```
Comment (rhc54) [inline src/tools/prted/prted.c:348]: Same here

**Review by rhc54 — CHANGES_REQUESTED** [review]
Couple of quick observations. Also, note that this effectively means that PRRTE v2.1 cannot be combined with any PMIx prior to v4.2 as the show-help subsystem will not work. Need to ponder that a bit to see how restrictive that becomes in reality - either we need a different solution for the release branch, or we have to change configure to enforce that requirement.

File: src/tools/prte/prte.c (hunk 1 above)
Comment (awlauria) [reply src/tools/prte/prte.c:468]: Right..this snuck through in the master PR. See https://github.com/openpmix/prrte/pull/1355

**Comment by awlauria** [comment]
> Couple of quick observations. Also, note that this effectively means that PRRTE v2.1 cannot be combined with any PMIx prior to v4.2 as the show-help subsystem will not work. Need to ponder that a bit to see how restrictive that becomes in reality - either we need a different solution for the release branch, or we have to change configure to enforce that requirement.

It easily cherry-picks over to the pmix v4.1 branch. I have not tested it, but it seems doable.

File: src/tools/prted/prted.c (hunk 2 above)
Comment (awlauria) [reply src/tools/prted/prted.c]: Resolved with latest commit.

**Comment by awlauria** [comment]
bot:ibm:xl:retest

**Comment by awlauria** [comment]
@rhc54 I added a commit that will allow this change to build with other PMIx versions which don't support AGG.

I definitely think a prrte with agg support + a PMIx without support should work/be supported, users will only lose agg support and logging messages will go to standard error. I could fairly easily add a configure time check for AGG support, and print a warning if it is not available, but I wouldn't go farther than that (ie failing to configure/compile).

> Couple of quick observations. Also, note that this effectively means that PRRTE v2.1 cannot be combined with any PMIx prior to v4.2 as the show-help subsystem will not work. Need to ponder that a bit to see how restrictive that becomes in reality - either we need a different solution for the release branch, or we have to change configure to enforce that requirement.

**Review by rhc54 — APPROVED** [review]

**Pull request merged**